index.rst

```diff
@@ -1,25 +1,429 @@
+I anticipate that our main use of default values will be for schema evolution: any field in the schema being read, but not the schema being written, will have the default value.
+
+Namespaces for Isolation
+------------------------
+
+It is important to be able to isolate subsystems from each other, both when running prerelease versions of software on the summit, and for unit testing anywhere.
+
+DDS has two mechanisms for isolation: domain ID and partition names.
+Domain ID is more complete isolation: messages using one domain ID cannot see messages on any other domain.
+Partition names offer less isolation: topics in all partitions must share the same schemas and quality of service settings.
+
+Kafka is simpler because one can run separate Kafka brokers to isolate systems.
+This is especially appropriate for running unit tests, as each package can spin up its own broker.
+However we also want to be able to isolate different versions of topics on the main Kafka system, e.g. in order to run an version of a controller.
+In order to support this, the Kafka verison of ts_salobj includes a namespace component in each topic name,
+where the value is specified by environment variable LSST_TOPIC_SUBNAME.
+
+Python Library
+--------------
+
+I used aiokafka to implement the Kafka version of ts_salobj, because it is widely used, has a high-level API, and supports asyncio.
+The obvious alternative is confluent_kafka, which is maintained by Confluent, the main company that provides commercial support for Kafka.
+Confluent_kafka is well supported and is designed to be fast.
+However, it has a rather clumsy API and has no native support for asyncio.
+I did test a prototype system that used confluent_kafka and found similar performance to aoikafka, probably because I was waiting for one ack when sending messages, which makes the broker the performance limitation, not the client library.
+Note that ts_salobj already uses confluent_kafka to specify topics, because aiokafka has no admin support.
+
+Optional Changes
+----------------
+
+I made several optional changes to Kafka ts_salobj, in an effort to clean up two minor SAL issues:
+
+* Renamed the SAL index field from *SAL_component_name*\ ID to ``private_index``.
+  This makes the field name match all other private fields, and makes it easier to use because one never needs to compute the name.
+
+* Removed the ``priority`` field from event topics, because we aren't using it.
+  This is a trivial change for ts_salobj, because it has no required use of the ``priority`` field:
+  one may set it, but as far as I can tell we have no code that does so.
+  ts_sal has ``priority`` as an explicit argument to a few event-related function calls.
+  If we agree to this change then I recommend that ts_sal continue to make the argument available, at least initially, for backwards compatibility, but make it optional, warn of non-zero values, and otherwise ignore it.
```

Comment:
> If we agree to adopt these change then it is recommended that ts_sal continue to make the argument available, at least initially, for backwards compatibility, but make it optional, warn of non-zero values, and otherwise ignore it.

Reply:
Good point. It is clumsy. I'm hoping this will do:

If we agree to adopt this change, we can preserve backwards compatibility by having ts_sal continue to make the argument available (at least initially) and ignore it, other than warning of non-zero values.

Reply:
This became irrelevant because RFCs were accepted that adopt these changes for DDS.